File: games/number-guesser/src/main/scala/org/jvmgames/scala/numberguesser/InputHandler.scala
package org.jvmgames.scala.numberguesser

import cats.effect.IO
import cats.effect.std.Console
import cats.syntax.all._

import org.jvmgames.scala.numberguesser.Domain.{GameDifficulty, Game}
import org.jvmgames.scala.numberguesser.GameLogic.isValidGuess
import org.jvmgames.scala.numberguesser.GameLogic.gameDifficultyFromIntSelection


object InputHandler:

  def welcomeMessage(difficulty: GameDifficulty): IO[Unit] =
    val welcomeMessage = s"""
    |Welcome to the Number Guesser Game!
    |
    |I am thinking of a number between 1 and ${difficulty.bound}.
    |Can you guess it within ${difficulty.maxGuesses} attempts?
    """.stripMargin
    Console[IO].println(welcomeMessage)

  def promptGameDifficulty: IO[GameDifficulty] =
    val prompt = """
    |Please select a difficulty level:
    |1. Easy (1-10, 5 guesses)
    |2. Medium (1-20, 4 guesses)
    |3. Hard (1-50, 3 guesses)
    |Enter the number corresponding to your choice:
    """.stripMargin
    Console[IO].println(prompt) *>
      Console[IO].readLine.flatMap { input =>
        gameDifficultyFromIntSelection(input) match
          case Some(difficulty) => IO.pure(difficulty)
          case None =>
            Console[IO].println("Invalid input. Please enter 1, 2, or 3.") *>
              promptGameDifficulty
      }
  
  def promptGuess(game: Game, gameDifficulty: GameDifficulty): IO[Int] =
    Console[IO].println("Enter your guess:") *>
      Console[IO].readLine.flatMap { input =>
        input.toIntOption match
          case Some(guess) if isValidGuess(guess, gameDifficulty) => IO.pure(guess)
          case Some(guess) =>
            Console[IO].println(s"Invalid guess. Please enter a number between 1 and ${gameDifficulty.bound}.") *>
              promptGuess(game, gameDifficulty)
          case None =>
            Console[IO].println("Invalid input. Please enter a valid integer.") *>
              promptGuess(game, gameDifficulty)
      }

  def hint(guess: Int, game: Game): IO[Unit] =
    val guessDiff = guess - game.secretNumber
    if guessDiff < 0 then
      Console[IO].println("Too low! Try again.")
    else
      Console[IO].println("Too high! Try again.")
  
  def displayRemainingGuesses(game: Game): IO[Unit] =
    Console[IO].println(s"Remaining guesses: ${game.remainingGuesses}")
  
  def congratulate(game: Game, gameDifficulty: GameDifficulty): IO[Unit] =
    Console[IO].println(s"Congratulations! You've guessed the number ${game.secretNumber}" +
      s" in ${gameDifficulty.maxGuesses - game.remainingGuesses} attempts!" + "\n" +
      s"⭐️⭐️⭐️⭐️⭐️⭐️⭐️⭐️⭐️⭐️⭐️⭐️⭐️⭐️⭐️⭐️⭐️⭐️⭐️⭐️")
  
  def sympathize(game: Game): IO[Unit] =
    Console[IO].println(s"Game over! You've used all your guesses. The number was ${game.secretNumber}" +
      s"😂🫵🏼😂🫵🏼😂🫵🏼😂🫵🏼😂🫵🏼")

  def promptPlayAgain: IO[Boolean] =
    Console[IO].println("Do you want to play again? (y/n)") *>
      Console[IO].readLine.flatMap {
        case "y" | "Y" => IO.pure(true)
        case "n" | "N" => IO.pure(false)
        case _ =>
          Console[IO].println("Invalid input. Please enter 'y' or 'n'.") *>
            promptPlayAgain
      }

File: games/number-guesser/src/main/java/org/jvmgames/java/numberguesser/InputHandler.java
package org.jvmgames.java.numberguesser;

import java.util.Scanner;
import java.util.Random;
import org.jvmgames.core.Game;
import org.jvmgames.java.numberguesser.GameDifficulty;


class InputHandler {

  public GameDifficulty getDifficulty(int choice) {
    return switch (choice) {
      case 1 -> GameDifficulty.EASY;
      case 2 -> GameDifficulty.MEDIUM;
      case 3 -> GameDifficulty.HARD;
      default -> null;
    };
  }

  public GameDifficulty promptDifficulty(Scanner scanner) {
    while (true) {
      System.out.println("""
        Please select a difficulty level:
        1. Easy (1-10, 5 guesses)
        2. Medium (1-20, 4 guesses)
        3. Hard (1-50, 3 guesses)
        Enter the number corresponding to your choice:
      """);
      int choice = scanner.nextInt();
      GameDifficulty difficulty = getDifficulty(choice);
      if (difficulty != null) {
        return difficulty;
      }
      System.out.println("Invalid choice. Please enter 1, 2, or 3.");
    }
  }

  public int handleGuess(Scanner scanner, GuessingGame game) {
    System.out.println("Enter your guess:");
    String guess = scanner.next();
    try {
      int num = Integer.parseInt(guess);
      if (num < 1 || num > game.getBound()) {
        System.out.println("Please enter a number between 1 and " + game.getBound() + ".");
        return handleGuess(scanner, game);
      }
      return num;
    } catch (NumberFormatException e) {
      System.out.println("Invalid input. Please enter a valid integer.");
      return handleGuess(scanner, game);
    }
  }

  public boolean promptPlayAgain(Scanner scanner) {
    System.out.println("Do you want to play again? (y/n)");
    String response = scanner.next().trim().toLowerCase();
    while (!response.equals("y") && !response.equals("n")) {
      System.out.println("Invalid input. Please enter 'y' for yes or 'n' for no.");
      response = scanner.next().trim().toLowerCase();
    }
    return response.equals("y");
  }
}
